File: 206-ReverseLinkedList/206-ReverseLinkedList.py
# Last updated: 6/25/2026, 9:15:27 AM

from sys import _current_frames# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        return self._reverseList(head,None)
    
    def _reverseList(self,head,prev):

        if not head:
            return prev
        
        ne=head.next
        head.next=prev
        return self._reverseList(ne,head)

File: 206-ReverseLinkedList/206-ReverseLinkedList.cpp
// Last updated: 6/25/2026, 9:15:32 AM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head == NULL) return head;
        if(head->next == NULL) return head;
        ListNode* prev; ListNode* trav = head; ListNode* n = trav->next;
        while(n!= NULL){
            if(trav == head) trav->next = NULL;
            else trav->next = prev;
            prev = trav;
            trav = n;
            n = n->next;
        }
        trav->next = prev;
        return trav;
    }
};